Orange/widgets/data/owpreprocess.py

```diff
@@ -535,6 +536,125 @@ def createinstance(params):
+        ibox = gui.indentedBox(widget=box)
+        p_spin = gui.hSlider(
+            widget=ibox, master=self, value="_p", minValue=0, maxValue=100,
+            step=4, ticks=20, labelFormat="%d %%", callback=self.update_p_spin,
+            disabled=self._strategy != RandomFeatureSelectEditor.Percentage)
+        self._spins[RandomFeatureSelectEditor.Percentage] = p_spin
+
+    def update_strategy(self, edited=True):
+        for strategy, spin in self._spins.items():
+            if self._strategy == strategy:
+                spin.setDisabled(False)
+            else:
+                spin.setDisabled(True)
+
+        if edited:
+            self.edited.emit()
+        else:
+            self.changed.emit()
+
+    def update_p_spin(self):
+        curr_val = self._p
+        prev_val = self._ticks_to_update_slider[1]
+        if (self._ticks_to_update_slider[0] is not None and (prev_val is None or
+            any([curr_val < tick <= prev_val or curr_val >= tick > prev_val
+                for tick in self._ticks_to_update_slider[0]]))):
+            self.edited.emit()
+        self._ticks_to_update_slider[1] = curr_val
+
+    def set_k(self, k):
+        if self._k != k:
+            self._k = k
+            self._spins[RandomFeatureSelectEditor.Fixed].setValue(self._k)
+            self.changed.emit()
+
+    def set_p(self, p):
+        if self._p != p:
+            self._p = p
+            self._spins[RandomFeatureSelectEditor.Percentage].setValue(self._p)
+            self.changed.emit()
+
+    def set_max_num_attributes(self, max_to_set):
+        if self._max_num_attributes != max_to_set:
+            self._max_num_attributes = max_to_set
+            k_spin = self._spins[RandomFeatureSelectEditor.Fixed]
+            p_spin = self._spins[RandomFeatureSelectEditor.Percentage]
+            if self._max_num_attributes is None:
+                k_spin.setMaximum(10)
+                p_spin.setTickInterval(100/5)
+                self._ticks_to_update_slider = [None, None]
+            else:
+                k_spin.setMaximum(self._max_num_attributes)
+                p_spin.setTickInterval(p_spin.maximum() /
+                                       self._max_num_attributes)
+                self._ticks_to_update_slider = \
+                    [[math.ceil(i*p_spin.maximum()/self._max_num_attributes)
+                      for i in range(1, self._max_num_attributes+1)], None]
+            self.changed.emit()
+
+    def set_strategy(self, strategy):
+        if self._strategy != strategy:
+            self._strategy = strategy
+            self.update_strategy(edited=False)
+
+    def setParameters(self, params):
+        self.set_k(params.get("k", 10))
+        self.set_p(params.get("p", 75))
+        self.set_max_num_attributes(params.get("max_num_attributes", None))
+        self.set_strategy(params.get("strategy",
+                                     RandomFeatureSelectEditor.Fixed))
```

Comment:
You can just assign to `self._k`, `self._p`, and `self._strategy` here, and their respective controls will be updated. That's what `master, value` parameters of gui.py controls are for. Then you probably also don't need `set_strategy()` and `update_strategy()` methods. If `parameters()` and `createinstance()` below are the only two entry points for `OWPreprocess` to use, you don't even need callbacks on your gui.py elements.